### PR TITLE
Speed up curve plotting

### LINKED ISSUE #61: Curve plotting in traces mode extremely slow
Plotting all views (raw, whitened, prediction and residual) in traces mode takes ~1s per update, which is way too slow.

- changes in the way curves are plotted / updated would make a difference
- `updateAutoRange()` in `update_plot()` takes a lot of time, hence might be worth disabling and using only manually after plotting
- `get_predicted_traces()` is really slow, speeding it up would definitely speed up both traces and colormap modes

### PR DESCRIPTION
**Pull request by shashwatsridhar**
Fixes #61.

- broke up `update_plot()` method in data_view_box.py by creating `_update_traces()` and `_update_colormap()` methods
- changed the way curves are plotted: `PlotCurveItem` objects are generated for number of channels requested, and the data held in each object is subsequently updated with every call to `_update_traces()` resulting in much faster plotting
- disabled auto range for the plot - `autoRange()` is now performed once at the end after plotting
- major speed up in `get_predicted_traces()`
  - used slightly faster variants of existing numpy commands
  - moved cupy calls outside function to make the function numba compatible
  - used numba to drastically speed up on-the-fly operation (results in a small delay the first time function is called)

A few quick benchmarks of the `update_plot()` method for a before (master branch) and after (current PR) comparison:

<table align="center">
<thead>
  <tr>
    <th rowspan="2">selected views</th>
    <th colspan="2">traces</th>
    <th colspan="2">colormap</th>
  </tr>
  <tr>
    <td>before</td>
    <td>after</td>
    <td>before</td>
    <td>after</td>
  </tr>
</thead>
<tbody>
  <tr>
    <td>raw</td>
    <td>200ms</td>
    <td>23ms</td>
    <td>4ms</td>
    <td>16ms</td>
  </tr>
  <tr>
    <td>whitened</td>
    <td>207ms</td>
    <td>30ms</td>
    <td>16ms</td>
    <td>40ms</td>
  </tr>
  <tr>
    <td>prediction</td>
    <td>216ms</td>
    <td>48ms</td>
    <td>48ms</td>
    <td>47ms</td>
  </tr>
  <tr>
    <td>residual</td>
    <td>229ms</td>
    <td>55ms</td>
    <td>43ms</td>
    <td>53ms</td>
  </tr>
  <tr>
    <td>all</td>
    <td>1.35s</td>
    <td>68ms</td>
    <td>-</td>
    <td>-</td>
  </tr>
</tbody>
</table>

There is significant improvement in curve plotting speed while sacrificing some speed in colormap plotting.

Additionally, the curve plotting speed is not adversely affected by the number of channels being plotted:

<table align="center">
<thead>
  <tr>
    <th rowspan="2">number<br>of<br>traces<br>(raw)</th>
    <th colspan="2">time</th>
  </tr>
  <tr>
    <td>before<br></td>
    <td>after</td>
  </tr>
</thead>
<tbody>
  <tr>
    <td>32</td>
    <td>158ms</td>
    <td>22ms</td>
  </tr>
  <tr>
    <td>62</td>
    <td>557ms</td>
    <td>31ms</td>
  </tr>
  <tr>
    <td>122</td>
    <td>2073ms</td>
    <td>50ms</td>
  </tr>
  <tr>
    <td>242</td>
    <td>7917ms</td>
    <td>80ms</td>
  </tr>
</tbody>
</table>

### CODE REVIEW

File: pykilosort/gui/data_view_box.py [hunk 1]
```diff
@@ -337,14 +351,20 @@ def change_displayed_channel_count(self, direction):
         if 0 < new_count <= total_channels:
             self.set_currently_displayed_channel_count(new_count)
             self.channelChanged.emit()
+            self.plot_item.clear()
+            self.create_plot_items()
             self.update_plot()
```
Comment (alexmorley) [inline pykilosort/gui/data_view_box.py]: we repeat these 4 lines together, do they have a name as a single operation? like
```
class DataViewBox
    ...
    def refresh(self):
        self.channelChanged.emit()
        self.plot_item.clear()
        self.create_plot_items()
        self.update_plot()
```

File: pykilosort/gui/data_view_box.py [hunk 2]
```diff
@@ -499,14 +521,103 @@ def generate_lookup_table(self, colormap_min, colormap_max, num_points=8192):
...
+                    if good_channels[channel]
+                    else self.bad_channel_color
+                )
+                curve.setPen(color=color, width=1)
+                curve.setData(
+                    traces.T[channel] *
+                    self.scale_factor *
+                    self.traces_scaling_factor[view]
+                )
+            except IndexError:
```
Comment (alexmorley) [inline pykilosort/gui/data_view_box.py:590]: suggestion: add a comment to note when this happens, if its not intended to happen then add a warning message

File: pykilosort/gui/sorter.py [hunk 3]
```diff
@@ -70,9 +71,15 @@ def get_whitened_traces(raw_data, probe, params, whitening_matrix):
     return whitened_traces, whitening_matrix
 
 
-def get_predicted_traces(matrix_U, matrix_W, sorting_result, time_limits):
-    W = cp.asnumpy(matrix_W)
-    U = cp.asnumpy(matrix_U)
+@jit(nopython=True)
```
Comment (alexmorley) [inline pykilosort/gui/sorter.py:74]: if we use numba for the first time do we need to add it to our requirements (conda & pip)

Comment (shashwatsridhar) [reply pykilosort/gui/sorter.py:74]: it looks like you already [added it](https://github.com/MouseLand/pykilosort/blame/a34edc8e657b6ee38147104c5726d034bb583c3d/pyks2.yml#L23) at some point last year, so we don't need to add it any more

File: pykilosort/gui/data_view_box.py (hunk 1 above)
Comment (shashwatsridhar) [reply pykilosort/gui/data_view_box.py]: good point
made a new function from them!

**Pull request merged**